### DISCUSSION TITLE
Using Distributed Mesh with UserObject prescribed properties

### DISCUSSION THREAD

**Discussion by lukuna** [idaholab/moose#21124] [Q&A Modules: General] [answered]
Hi All,
I am trying to run some simulations using an application built on the MOOSE framework (specifically: https://github.com/ngrilli/c_pfor_am).  It works wonderfully for my purposes but I have come across a strange issue.  

I am trying to run some fairly large calculations (500K+ Voxels), and am seeing a lot of overhead when I run on many nodes.  To clarify, the same input file running 250 cores runs much faster on a cluster where I have 126 cores/node vs 48 cores/node.  To the point where running my calculation on more than 2 nodes actually slows it down (!!!).  In an effort to remedy this I looked to try to use the `--distributed-mesh` option and have come across a strange issue.  Before I get to the issue let me introduce why I think it may be happening.  In these calculations UserObjects are used to read in microstructure (Euler angle) and temperature data. 
 I believe the images below clearly show the issue:  

Distributed:
![Distributed](https://user-images.githubusercontent.com/94388491/170135022-42bef607-347e-450e-9bf2-1e04245adf71.jpeg)

Not Distributed:
![Not_Distributed](https://user-images.githubusercontent.com/94388491/170135024-647978f3-01f7-44ce-8611-2a928d5e5aa6.jpeg)
 
The way that the temperature is being read onto the mesh seems be messed up by the distributed mesh.  My mesh is as follows:
```
  [./gen]
    type = GeneratedMeshGenerator
    dim = 3
    nx = 40
    ny = 46
    nz = 18
    xmax = 240.0
    ymax = 276.0
    zmax = 108.0
    elem_type = HEX8
    displacements = 'disp_x disp_y disp_z'
  [../]
```
and I am using the two UserObjects located in the c_pfor_am app:
  ```
[./prop_read]
    type = GrainPropertyReadFile
    prop_file_name = 'CA3DVoxSizeTestLargeAngles.txt'
    # Enter file data as prop#1, prop#2, .., prop#nprop
    nprop = 3
    read_type = element
  [../]

  [./temperature_read]
    type = LaserTempReadFile
    temperature_file_name = 'VoxSizeTestLargeTemp.txt'
    temperature_num_step = 7
  [../]
```
Found [here](https://github.com/ngrilli/c_pfor_am/blob/main/src/userobjects/GrainPropertyReadFile.C) and [here](https://github.com/ngrilli/c_pfor_am/blob/main/src/userobjects/LaserTempReadFile.C).  Does anybody have any ideas or suggestions as to how this can be remedied?  Is this something internal with MOOSE, or just a slight adjustment that needs to be made within the app?  

Thanks for your time,
Lukasz Kuna

The Executioner I am using for anyone curious regarding the HPC slowdown is:

```

[Executioner]
  type = Transient
  solve_type = 'NEWTON'
  petsc_options_iname = '-pc_hypre_type'
  petsc_options_value = 'boomerang'
  line_search = 'default'
  automatic_scaling = true

  l_max_its = 30
  nl_max_its = 15
  nl_rel_tol = 1e-3
  nl_abs_tol = 1e-3
  l_tol = 1e-5

  [./TimeStepper]
    type = FunctionDTGrowth
    function = dts
    cutback_factor_at_failure = 0.2
    growth_factor = 1.5
  [../]

  start_time = 0.0
  end_time = 2800
  dtmin = 1.0e-99
  timestep_tolerance = 1.0e-99

[]

```

I had used `lu` before, but it was maxing out the memory on the HPC nodes I was using so I switched to `hypre`

Any suggestions regarding this would also be greatly appreciated.

**Comment by GiudGiud** [comment, marked as answer]
Hello

you need to turn off mesh renumbering to use this user object with a distrivuted mesh. Don’t you get a warning?
or did you do this already?

guillaume

**Reply by GiudGiud** [reply]
allow_mesh_renumbering = false in Mesh block

**Reply by lukuna** [reply]
This goes inside the [Mesh] block or inside of [./gen]?
When placed in [Mesh] I am seeing:

`/Users/lukasz/projects/c_pfor_am/VoxSizeTestLargeWall.i:6: unused parameter 'Mesh/allow_mesh_renumbering'`

and obviously no difference in the resulting mesh.

**Reply by GiudGiud** [reply]
It s just allow_renumbering it turns out. I had it wrong

**Reply by lukuna** [reply]
Thank you Guillaume.  This has solved my problem.  It would be very useful if this was documented along with the Mesh System (or wherever the distributed mesh documentation is).

**Reply by GiudGiud** [reply]
This is more of a problem with the user object you are using to map the CSV file to the mesh. I ll add it to the user object documentation

**Comment by lukuna** [comment]
@ngrilli This is the solution to the distributed-mesh issue I was having.  Just FYI.
